### ISSUE TITLE
Finetune Error

### ISSUE THREAD

**Issue by nubitz** [DarioFT/ComfyUI-Qwen3-TTS#6]
I like to first say thank you for putting this together. Below is the error I am getting when using the following node.

Qwen3FineTune
num_samples should be a positive integer value, but got num_samples=0

Both json files dataset.jsonl and dataset_codes.jsonl  are empty.

**Comment by rekuenkdr** [comment]
> > > Stupid question, but how can I add the new finetune build into comfyui? I'd like to modify the options provided in the pull [#14](https://github.com/DarioFT/ComfyUI-Qwen3-TTS/pull/14)
> > 
> > 
> > Loading the finetune model? just use the model path to the checkpoint on the TTS Custom Voice node. I uploaded a workflow, check the examples folder. (The one with the Benchmark does just that).
> > 1st block finetunes, 2nd generates, just modify the path and disable the finetune group.
> > Or what do you need by modify? Instruct does not work very well with the finetunes models but you can try it on the same TTS Custom Voice node.
> > I can modify a node to list the checkpoints on a dropdown to make it easier, I didn't bother much because was for testing.
> > If it is the last PR changes that you are not seeing just update this nodes on the Comfy Manager.
> > I have also a few more changes implemented to help with the finetune (output lr to rate to the console to see when the model converges and stops learning), and I have also modify the finetune node to save the checkpoints every epoch/save every X steps in replacement for the current logic ( that deletes the intermediate checkpoints if you don't mark to keep them).
> > I will push the changes tomorrow if I have the time.
> 
> I was referring to your node with the additional fine tune parameters you added in the pull release. I installed it today in ComfyUI and didn't see your options so was checking if they were merged in the main build or if I needed to install/update the node manually. The main thing I need is the batching you added for training. Some of my datasets are several minutes long which was causing OOM errors and while I could manually splice them, wanted to see if your batching helps with that.
> 
> I appreciate your time and replying.

Sorry I misunderstood, just go to your custom_nodes/ComfyUI-Qwen3-TTS/ folder and git pull the latest changes if you are not still seeing them. Maybe @DarioFT can do a release later with proper version tags so ComfyUI know when there is an update.

**Comment by rekuenkdr** [comment]
Qwen have just publish a new recommended settings for stable finetune, in case anyone have the time to give them a try.
https://github.com/QwenLM/Qwen3-TTS/commit/680d4e9d0ef904c1673062621059c2e42c7033fb

**Comment by vytskalt** [comment]
The training loss dropped to 8 instead of stagnating at 12 for me. But from the first couple generations I don't really hear much of a difference in emotion

**Comment by rekuenkdr** [comment]
Did you update qwen_tts library before trying?

Is going to be slower to train now, but let's give it a try for those 10 epocs and compare. 

If there's no difference you can stick with the tested params you already have.

**Comment by vytskalt** [comment]
> Did you update qwen_tts library before trying?
> 
> Is going to be slower to train now, but let's give it a try for those 10 epocs and compare.
> 
> If there's no difference you can stick with the tested params you already have.

I applied the 0.3 multiplier patch manually

**Comment by rekuenkdr** [comment]
Can anyone also try this PR? it seems to have found the culprit to the speed going nuts on finetuning:
https://github.com/QwenLM/Qwen3-TTS/pull/178

**Comment by vytskalt** [comment]
The speed is indeed better for me

**Comment by rekuenkdr** [comment]
I have fixed too the same problem with inference on voice clone with streaming, finally, and merge that PR for the finetune fix. 
I have implemented it on a custom qwen_tts library that have streaming pcm enabled, for an AI assistant project I'm working on, and I can confirm it is fixed on inference. https://github.com/rekuenkdr/Qwen3-TTS-streaming

I will test the finetune also as soon as I can, I'm glad is working for you @vytskalt

**Comment by rekuenkdr** [comment]
> Can anyone also try this PR? it seems to have found the culprit to the speed going nuts on finetuning: [QwenLM/Qwen3-TTS#178](https://github.com/QwenLM/Qwen3-TTS/pull/178)

I have reverted that PR, and use instead this solution:
https://github.com/QwenLM/Qwen3-TTS/issues/179#issuecomment-3870059313

Seems to improve the result, but still getting speed issues on finetuning. default batch and lr 20 epochs. will train a little more.

**Comment by TarekWSw** [comment]
@rekuenkdr 
This is training script i am using and those are the configs, is this combination should yield are good results, or i should extend the training epochs? I am fine-tuning for the arabic langauge. My dataset is around 103h (33k samples of audio segments, ranging between 3-16s)
Also is it better to modify the loss in `qwen_tts/core/models/modeling_qwen3_tts.py, line 1243:` as mentioned here https://github.com/QwenLM/Qwen3-TTS/issues/179#issuecomment-3870059313
```
  batch_size: 4
  learning_rate: 2e-6
  weight_decay: 0.01
  num_epochs: 5
  gradient_accumulation_steps: 8
  mixed_precision: "bf16"
  max_grad_norm: 1.0
  sub_talker_loss_weight: 0.3
```

```
import ....

target_speaker_embedding = None

def train(cfg: SFTConfig) -> None:
    """Run supervised fine-tuning on a Qwen3-TTS base model."""
    global target_speaker_embedding

    accelerator = Accelerator(
        gradient_accumulation_steps=cfg.training.gradient_accumulation_steps,
        mixed_precision=cfg.training.mixed_precision,
        log_with="tensorboard",
        project_dir=cfg.output.logs_dir,
    )

    qwen3tts = Qwen3TTSModel.from_pretrained(
        cfg.model.init_model_path,
        torch_dtype=torch.bfloat16,
        attn_implementation=cfg.model.attn_implementation,
    )
    config = AutoConfig.from_pretrained(cfg.model.init_model_path)

    with open(cfg.data.train_jsonl) as f:
        train_data = [json.loads(line) for line in f]

    dataset = TTSDataset(
        train_data, qwen3tts.processor, config, jsonl_path=cfg.data.train_jsonl
    )
    train_dataloader = DataLoader(
        dataset,
        batch_size=cfg.training.batch_size,
        shuffle=True,
        collate_fn=dataset.collate_fn,
    )

    optimizer = AdamW(
        qwen3tts.model.parameters(),
        lr=cfg.training.learning_rate,
        weight_decay=cfg.training.weight_decay,
    )

    model, optimizer, train_dataloader = accelerator.prepare(
        qwen3tts.model, optimizer, train_dataloader
    )

    model.train()

    for epoch in range(cfg.training.num_epochs):
        for step, batch in enumerate(train_dataloader):
            with accelerator.accumulate(model):
                input_ids = batch["input_ids"]
                codec_ids = batch["codec_ids"]
                ref_mels = batch["ref_mels"]
                text_embedding_mask = batch["text_embedding_mask"]
                codec_embedding_mask = batch["codec_embedding_mask"]
                attention_mask = batch["attention_mask"]
                codec_0_labels = batch["codec_0_labels"]
                codec_mask = batch["codec_mask"]

                speaker_embedding = model.speaker_encoder(
                    ref_mels.to(model.device).to(model.dtype)
                ).detach()
                if target_speaker_embedding is None:
                    target_speaker_embedding = speaker_embedding

                input_text_ids = input_ids[:, :, 0]
                input_codec_ids = input_ids[:, :, 1]

                input_text_embedding = (
                    model.talker.model.text_embedding(input_text_ids)
                    * text_embedding_mask
                )
                input_codec_embedding = (
                    model.talker.model.codec_embedding(input_codec_ids)
                    * codec_embedding_mask
                )
                input_codec_embedding[:, 6, :] = speaker_embedding

                input_embeddings = input_text_embedding + input_codec_embedding

                outputs = model.talker(
                    inputs_embeds=input_embeddings,
                    attention_mask=attention_mask,
                    labels=codec_0_labels,
                    output_hidden_states=True,
                )

                hidden_states = outputs.hidden_states[0][-1]
                target_codec_mask = codec_mask[:, 1:]
                talker_hidden_states = hidden_states[:, :-1][target_codec_mask]
                talker_codec_ids = codec_ids[:, 1:][target_codec_mask]

                _, sub_talker_loss = model.talker.forward_sub_talker_finetune(
                    talker_codec_ids, talker_hidden_states
                )

                loss = (
                    outputs.loss + cfg.training.sub_talker_loss_weight * sub_talker_loss
                )

                accelerator.backward(loss)

                if accelerator.sync_gradients:
                    accelerator.clip_grad_norm_(
                        model.parameters(), cfg.training.max_grad_norm
                    )

                optimizer.step()
                optimizer.zero_grad()

            if step % 10 == 0:
                accelerator.print(
                    f"Epoch {epoch} | Step {step} | Loss: {loss.item():.4f}"
                )

        if accelerator.is_main_process:
            output_dir = os.path.join(
                cfg.output.model_path, f"checkpoint-epoch-{epoch}"
            )
            shutil.copytree(cfg.model.init_model_path, output_dir, dirs_exist_ok=True)

            input_config_file = os.path.join(cfg.model.init_model_path, "config.json")
            output_config_file = os.path.join(output_dir, "config.json")
            with open(input_config_file, encoding="utf-8") as f:
                config_dict = json.load(f)
            config_dict["tts_model_type"] = "custom_voice"
            talker_config = config_dict.get("talker_config", {})
            talker_config["spk_id"] = {cfg.speaker.name: 3000}
            talker_config["spk_is_dialect"] = {cfg.speaker.name: False}
            config_dict["talker_config"] = talker_config

            with open(output_config_file, "w", encoding="utf-8") as f:
                json.dump(config_dict, f, indent=2, ensure_ascii=False)

            unwrapped_model = accelerator.unwrap_model(model)
            state_dict = {
                k: v.detach().to("cpu") for k, v in unwrapped_model.state_dict().items()
            }

            drop_prefix = "speaker_encoder"
            keys_to_drop = [k for k in state_dict if k.startswith(drop_prefix)]
            for k in keys_to_drop:
                del state_dict[k]

            weight = state_dict["talker.model.codec_embedding.weight"]
            state_dict["talker.model.codec_embedding.weight"][3000] = (
                target_speaker_embedding[0].detach().to(weight.device).to(weight.dtype)
            )
            save_path = os.path.join(output_dir, "model.safetensors")
            save_file(state_dict, save_path)

```

**Comment by rekuenkdr** [comment]
> [@rekuenkdr](https://github.com/rekuenkdr) This is training script i am using and those are the configs, is this combination should yield are good results, or i should extend the training epochs? I am fine-tuning for the arabic langauge. My dataset is around 103h (33k samples of audio segments, ranging between 3-16s) Also is it better to modify the loss in `qwen_tts/core/models/modeling_qwen3_tts.py, line 1243:` as mentioned here [QwenLM/Qwen3-TTS#179 (comment)](https://github.com/QwenLM/Qwen3-TTS/issues/179#issuecomment-3870059313)
> 
> ```
>   batch_size: 4
>   learning_rate: 2e-6
>   weight_decay: 0.01
>   num_epochs: 5
>   gradient_accumulation_steps: 8
>   mixed_precision: "bf16"
>   max_grad_norm: 1.0
>   sub_talker_loss_weight: 0.3
> ```
> 
> ```
> import ....
> 
> target_speaker_embedding = None
> 
> def train(cfg: SFTConfig) -> None:
>     """Run supervised fine-tuning on a Qwen3-TTS base model."""
>     global target_speaker_embedding
> 
>     accelerator = Accelerator(
>         gradient_accumulation_steps=cfg.training.gradient_accumulation_steps,
>         mixed_precision=cfg.training.mixed_precision,
>         log_with="tensorboard",
>         project_dir=cfg.output.logs_dir,
>     )
> 
>     qwen3tts = Qwen3TTSModel.from_pretrained(
>         cfg.model.init_model_path,
>         torch_dtype=torch.bfloat16,
>         attn_implementation=cfg.model.attn_implementation,
>     )
>     config = AutoConfig.from_pretrained(cfg.model.init_model_path)
> 
>     with open(cfg.data.train_jsonl) as f:
>         train_data = [json.loads(line) for line in f]
> 
>     dataset = TTSDataset(
>         train_data, qwen3tts.processor, config, jsonl_path=cfg.data.train_jsonl
>     )
>     train_dataloader = DataLoader(
>         dataset,
>         batch_size=cfg.training.batch_size,
>         shuffle=True,
>         collate_fn=dataset.collate_fn,
>     )
> 
>     optimizer = AdamW(
>         qwen3tts.model.parameters(),
>         lr=cfg.training.learning_rate,
>         weight_decay=cfg.training.weight_decay,
>     )
> 
>     model, optimizer, train_dataloader = accelerator.prepare(
>         qwen3tts.model, optimizer, train_dataloader
>     )
> 
>     model.train()
> 
>     for epoch in range(cfg.training.num_epochs):
>         for step, batch in enumerate(train_dataloader):
>             with accelerator.accumulate(model):
>                 input_ids = batch["input_ids"]
>                 codec_ids = batch["codec_ids"]
>                 ref_mels = batch["ref_mels"]
>                 text_embedding_mask = batch["text_embedding_mask"]
>                 codec_embedding_mask = batch["codec_embedding_mask"]
>                 attention_mask = batch["attention_mask"]
>                 codec_0_labels = batch["codec_0_labels"]
>                 codec_mask = batch["codec_mask"]
> 
>                 speaker_embedding = model.speaker_encoder(
>                     ref_mels.to(model.device).to(model.dtype)
>                 ).detach()
>                 if target_speaker_embedding is None:
>                     target_speaker_embedding = speaker_embedding
> 
>                 input_text_ids = input_ids[:, :, 0]
>                 input_codec_ids = input_ids[:, :, 1]
> 
>                 input_text_embedding = (
>                     model.talker.model.text_embedding(input_text_ids)
>                     * text_embedding_mask
>                 )
>                 input_codec_embedding = (
>                     model.talker.model.codec_embedding(input_codec_ids)
>                     * codec_embedding_mask
>                 )
>                 input_codec_embedding[:, 6, :] = speaker_embedding
> 
>                 input_embeddings = input_text_embedding + input_codec_embedding
> 
>                 outputs = model.talker(
>                     inputs_embeds=input_embeddings,
>                     attention_mask=attention_mask,
>                     labels=codec_0_labels,
>                     output_hidden_states=True,
>                 )
> 
>                 hidden_states = outputs.hidden_states[0][-1]
>                 target_codec_mask = codec_mask[:, 1:]
>                 talker_hidden_states = hidden_states[:, :-1][target_codec_mask]
>                 talker_codec_ids = codec_ids[:, 1:][target_codec_mask]
> 
>                 _, sub_talker_loss = model.talker.forward_sub_talker_finetune(
>                     talker_codec_ids, talker_hidden_states
>                 )
> 
>                 loss = (
>                     outputs.loss + cfg.training.sub_talker_loss_weight * sub_talker_loss
>                 )
> 
>                 accelerator.backward(loss)
> 
>                 if accelerator.sync_gradients:
>                     accelerator.clip_grad_norm_(
>                         model.parameters(), cfg.training.max_grad_norm
>                     )
> 
>                 optimizer.step()
>                 optimizer.zero_grad()
> 
>             if step % 10 == 0:
>                 accelerator.print(
>                     f"Epoch {epoch} | Step {step} | Loss: {loss.item():.4f}"
>                 )
> 
>         if accelerator.is_main_process:
>             output_dir = os.path.join(
>                 cfg.output.model_path, f"checkpoint-epoch-{epoch}"
>             )
>             shutil.copytree(cfg.model.init_model_path, output_dir, dirs_exist_ok=True)
> 
>             input_config_file = os.path.join(cfg.model.init_model_path, "config.json")
>             output_config_file = os.path.join(output_dir, "config.json")
>             with open(input_config_file, encoding="utf-8") as f:
>                 config_dict = json.load(f)
>             config_dict["tts_model_type"] = "custom_voice"
>             talker_config = config_dict.get("talker_config", {})
>             talker_config["spk_id"] = {cfg.speaker.name: 3000}
>             talker_config["spk_is_dialect"] = {cfg.speaker.name: False}
>             config_dict["talker_config"] = talker_config
> 
>             with open(output_config_file, "w", encoding="utf-8") as f:
>                 json.dump(config_dict, f, indent=2, ensure_ascii=False)
> 
>             unwrapped_model = accelerator.unwrap_model(model)
>             state_dict = {
>                 k: v.detach().to("cpu") for k, v in unwrapped_model.state_dict().items()
>             }
> 
>             drop_prefix = "speaker_encoder"
>             keys_to_drop = [k for k in state_dict if k.startswith(drop_prefix)]
>             for k in keys_to_drop:
>                 del state_dict[k]
> 
>             weight = state_dict["talker.model.codec_embedding.weight"]
>             state_dict["talker.model.codec_embedding.weight"][3000] = (
>                 target_speaker_embedding[0].detach().to(weight.device).to(weight.dtype)
>             )
>             save_path = os.path.join(output_dir, "model.safetensors")
>             save_file(state_dict, save_path)
> ```

The main Qwen updated the recommended epochs to 10, with your settings (effective batch 32, what you already have), learning a new language that is not supported will take more time and it will also depend on the dataset size.

I haven't tried to finetune a new language, maybe you could asked directly on Qwen's official repository, there are more people there trying to do just the same.
